axelor-production/src/main/java/com/axelor/apps/production/service/BillOfMaterialServiceImpl.java

```diff
@@ -300,4 +303,59 @@ public void removeInvalidTree(List<Long> validBomIds, BillOfMaterial bom) {
   public void setBillOfMaterialAsDefault(BillOfMaterial billOfMaterial) {
     billOfMaterial.getProduct().setDefaultBillOfMaterial(billOfMaterial);
   }
+
+  @Override
+  public String computeName(BillOfMaterial bom) {
+    return bom.getProduct().getName()
+        + " - "
+        + bom.getQty()
+        + bom.getUnit().getName()
+        + " - "
+        + bom.getId();
+  }
+
+  @Override
+  @Transactional
+  public Set<BillOfMaterial> addRawMaterials(
+      BillOfMaterial billOfMaterial, ArrayList<LinkedHashMap<String, Object>> rawMaterials) {
+    Set<BillOfMaterial> components = billOfMaterial.getBillOfMaterialSet();
+    int priority =
+        Collections.max(
+            billOfMaterial
+                .getBillOfMaterialSet()
+                .stream()
+                .map(it -> it.getPriority())
+                .collect(Collectors.toSet()));
+
+    for (LinkedHashMap<String, Object> rawMaterial : rawMaterials) {
+      priority += 10;
+      BillOfMaterial newComponent =
+          createBomFromRawMaterial(Long.valueOf((int) rawMaterial.get("id")), priority);
+      components.add(newComponent);
+    }
+
+    return components;
+  }
+
+  private BillOfMaterial createBomFromRawMaterial(long productId, int priority) {
+    BillOfMaterial newBom = new BillOfMaterial();
+    JPA.runInTransaction(
```

Comment:
Transaction not required here because you return the components to the view. Moreover, BillOfmaterial change are not saved before run process